text/0076-entity-slice-validation.md

```diff
@@ -0,0 +1,440 @@
+    // This will be our entity slice
+   let mut es : Map<EUID, EntityData> = {};
+
+   let mut worklist : Set<EUID> = {principal, action, resource, all root euids in context};
+   for euid in worklist {
+     es.insert(euid, get-entity(euid));
+   }
+   
+   let current-level = 0;
+   
+   while current-level <= level || worklist.empty? {
+        // `t` will become our new worklist
+        let t : Set<EUID> = {};
+        // Clear our worklist
+        for euid in worklist {
+            if !es.contains(euid) {
+                es.insert(euid, get-entity(euid));
+            }
+            ; Set of all euids mentioned in this entities data
+            let new-euids = filter(not-in-es?, all-euids(es.get(euid));
+            t = t.union(new-euids);
+        }
+        // set t to be the new worklist
+        s = t;
+        // increment level
+        level = level + 1;
+   }
+
+    return es;
+}
+
+```
+
+### Soundness
+
+Leveled types permit us to generalize the soundness statement we use today. Currently, soundness assumes that expressions like `principal.manager` will never error, i.e., that the proper entity information is available. Now we assume dereferences are only available up to a certain level. In particular, we define a well-formedness judgment μ⊢v:τwhere μ is an entity store, v is a value, and τ is a possibly leveled type. This judgment is standard for values having non-entity type, e.g.,
+
+```
+mu |- i : Long
+
+mu |- s : String
+
+mu |- v1 : T1 ... mu |- vn : Tn 
+-------------------------------------------------------
+mu |- { f1: v1, ..., fn: vn } : { f1: T1, ..., fn: Tn }
+```
+
+There are two rules for entity literals of type `E::s`. The first says level `0` is always allowed. The second says a literal can have a level `n` only if `n` dereferences are possible starting from `E::s`:
+
+```
+mu |- E::s : E(0)
+
+mu(E::s) = (v,h) where v = { f1: v1, ..., fn: vn } and h is an ancestor set
+mu |- v1 : T1 ... mu |- vn : Tn
+where level(T1) >= n-1
+------------------------------------------------------------------------------------
+mu |- E::s : E(n) where n > 0
+```
+
+The auxiliary function `level(T)` identifies the lower-bound of the level of type `T`. It’s defined as follows:
+
+* `level(E(n))` = `n`
+* `level({ f1: T1, ..., fn: Tn })`  = `m` where `m` = `min(level(T1),...,level(Tn))`
+* `level(Long)` = `level(String)` = `level(Boolean)` = `level(Set(T))` = (etc.) = ∞ for all other types
+
+Thus the soundness statement says that 
+
+1. If policy set `p` is validated against a schema `s` with level `l`, where  `l < ∞` .
+2. If entity store μ is validated against schema `s` 
+3. If request σ is validated against schema `s` 
+4. If μ ⊢ σ(principal): E(l) where E is the expected type of the principal for this particular action, and  similarly check the proper leveling of action, request, and context.
+    1. We prove separately that `slice` and σ produces a store μ that satisfies this property
```

Comment:
I was a little confused about how this sub-point is relevant. Maybe elaborate more, or delete it?

Reply:
As a fix, I pulled out the soundness statement for `slice` on its own, below.